### PR TITLE
feat: handle voided transactions

### PR DESCRIPTION
**Pull request by andreabadesso**

### CODE REVIEW

File: src/commons.ts [hunk 1]
```diff
@@ -378,6 +378,55 @@ export const handleVoidedTxList = async (mysql: ServerlessMysql, txs: Tx[]): Pro
   return [newTxs, affectedUtxoList];
 };
 
+/**
+ * Handles a reorg by finding the last valid block on the service's database and
```
Comment (pedroferreira1) [inline src/commons.ts]: The docstring (including parameters and return) is wrong

File: src/commons.ts [hunk 2]
```diff
@@ -378,6 +378,55 @@ export const handleVoidedTxList = async (mysql: ServerlessMysql, txs: Tx[]): Pro
...
+    txs = newTxs;
+    affectedUtxoList = [...affectedUtxoList, ...newAffectedUtxoList];
+  }
+
+  // fetch all addresses affected by the voided tx
+  const affectedAddresses = affectedUtxoList.reduce((acc: Set<string>, utxo: DbTxOutput) => acc.add(utxo.address), new Set<string>());
+
+  console.log(`Rebuilding balances for ${affectedAddresses.size} addresses.`);
+  if (affectedAddresses.size > 0) {
+    const addresses = [...affectedAddresses];
```
Comment (pedroferreira1) [inline src/commons.ts:406]: Why do you recreate the list again here?

File: tests/txProcessor.test.ts [hunk 3]
```diff
@@ -266,3 +267,69 @@ test('receive token creation tx', async () => {
...
+  // void the transaction
+  await txProcessor.handleVoidedTx(tx);
+
+  // both utxos should be voided
+  await expect(checkUtxoTable(mysql, 3, txId2, 0, token, addr, 2000, 0, null, null, false, null, true)).resolves.toBe(true);
+  await expect(checkUtxoTable(mysql, 3, txId2, 1, token, 'other', 500, 0, null, null, false, null, true)).resolves.toBe(true);
+  // the original utxo should not be voided and should not have been spent
+  await expect(checkUtxoTable(mysql, 3, txId1, 0, token, addr, 2500, 0, null, null, false, null, false)).resolves.toBe(true);
+
+  await expect(checkAddressBalanceTable(mysql, 1, addr, token, 2500, 0, 0, 1)).resolves.toBe(true);
```
Comment (pedroferreira1) [inline tests/txProcessor.test.ts:350]: Maybe we should add a tx3 in this test that spends tx2. With tx2 as voided, tx3 should become voided as well and tx1 unspent. What do you think?

**Review by pedroferreira1 — CHANGES_REQUESTED** [review]

File: src/commons.ts (hunk 2 above)
Comment (andreabadesso) [reply src/commons.ts:406]: `affectedAddresses` is a `Set`, I need to convert it to an `array`

Comment (msbrogli) [inline src/commons.ts:406]: Maybe log which address is being recalculated?

**Review by msbrogli — APPROVED** [review]

File: src/commons.ts (hunk 2 above)
Comment (andreabadesso) [reply src/commons.ts:406]: I agree, the list probably won't be large so we can safely log

Done in d38c66bf0d8304c4d694e4bf499c4b17300dc994

File: src/commons.ts (hunk 1 above)
Comment (andreabadesso) [reply src/commons.ts]: Done in c6fa85e97944dde34ef931a0d8d50cb516b825df

File: tests/txProcessor.test.ts (hunk 3 above)
Comment (andreabadesso) [reply tests/txProcessor.test.ts:350]: I agree!

Done in 27d47abafca129b3c2b22d03757262eb18f44208

**Review by pedroferreira1 — APPROVED** [review]

**Pull request merged**